rpc/item.c

```diff
@@ -566,15 +623,31 @@ M0_INTERNAL bool m0_rpc_item_xid_check(struct m0_rpc_item *item,
 	if (!rpc_item_needs_xid(item))
 		return true;
 
-	M0_LOG(M0_DEBUG, "item: "ITEM_FMT" session=%p xid=%"PRIu64
-	       " s_xid=%"PRIu64, ITEM_ARG(item), sess, xid, sess->s_xid);
+	M0_LOG(M0_DEBUG, "item: "ITEM_FMT" session=%p osr_xid=%"PRIu64" "
+	       "osr_session_xid_min=%"PRIu64" s_xid=%"PRIu64,
+	       ITEM_ARG(item), sess, item->ri_header.osr_session_xid_min,
+	       xid, sess->s_xid);
 	/*
 	 * Purge cache on every N-th packet
 	 * (not on every one - that could be pretty expensive).
 	 */
 	if ((xid & 0xff) == 0)
 		m0_rpc_item_cache_purge(&sess->s_reply_cache);
 
+	/*
+	 * Check if an item had been cancelled on the other side.
+	 * If it had, sess->s_xid has to be updated, otherwise this rpc session
+	 * will wait indefinitely for the item, which is not going to come.
+	 */
```

Comment:
More explanations needed here to help understanding.
1. If an rpc item has been cancelled, the server will not process it any longer, and no reply will be sent back to client.
2. Can you explain, in this case, why is the condition (sess->s_xid < item->ri_header.osr_session_xid_min - 1) is true?

3. According to my understanding, the "item->ri_header.osr_session_xid_min" is the minimal active xid on sender side when there are multiple pending items. 
    I don't understand why the "sess->s_xid" would be possibly less than it.

4. The "item->ri_header.osr_session_xid_min" is now sent to receiver/other side. Could the other side do something with it?
     For example, reply item would piggy back the minimal xid on the other side?
     So the sender side can know that some of the items have been cancelled and no reply would come. (Also PING fop reply can carry such messages, too).

Reply:
1. I didn't get this. Cancellation happens on the other side, server couldn't know about cancellation without this patch.

2. Because the next item, which is expected to come, has to have xid = s_xid + 1. If it's cancelled then it will get finalised, and then it will no longer be any item with xid = s_xid + 1 or less on the other side. osr_session_xid_min will get updated with the minimum xid and s_xid will be adjusted.

3. If you enqueue 1000 items and then cancel first 999 of them, then you will have s_xid much less than osr_session_xid_min.

4. Yes, it could. What kind of issues this may solve?